File: api_gateway/src/models/messages.py
"""
Data models for the API Gateway.

Defines:
- Pydantic models for API validation
- Enums for message types and roles
- Support for A2A protocol features
"""

from datetime import datetime
from typing import List, Optional, Dict, Any
from enum import Enum
from pydantic import BaseModel, Field

# Import standardized message schema
from .message_schema import (
    MessageType as SocketMessageType,
    BaseMessage as SocketBaseMessage,
    Message as SocketMessage,
    UserTextMessage,
    AgentTextMessage
)

# --- Legacy Enums (for database compatibility) ---

class MessageType(str, Enum):
    """Types of messages in the system (DB storage format)."""
    USER = "user"
    AGENT = "agent"
    SYSTEM = "system"
    ERROR = "error"
    TASK = "task"  # For A2A task-related messages

class MessageRole(str, Enum):
    """Roles for messages in A2A communication."""
    INITIATOR = "initiator"
    RESPONDER = "responder"
    OBSERVER = "observer"
    COORDINATOR = "coordinator"

# --- API Models ---

class ErrorResponse(BaseModel):
    """Standard error response format."""
    code: int = Field(..., description="Error code")
    message: str = Field(..., description="Error message")
    details: Optional[Dict[str, Any]] = Field(default=None, description="Additional error details")

class Message(BaseModel):
    """Model for a message in database storage."""
    id: Optional[int] = None
    message_uuid: str
    session_id: str
    type: MessageType
    role: Optional[MessageRole] = None
    agent_id: Optional[str] = None
    parts: List[Dict[str, Any]]
    message_metadata: Optional[Dict[str, Any]] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    in_reply_to: Optional[str] = None
    context_refs: Optional[List[Dict[str, Any]]] = None
    capabilities_used: Optional[List[str]] = None

    class Config:
        from_attributes = True

    def to_socket_message(self) -> SocketMessage:
        """Convert database message to socket message format."""
        # Extract content from parts
        content = ""
        if self.parts:
            for part in self.parts:
                if part.get("type") == "text":
                    content += part.get("content", "")
        
        # Build common base message fields
        base_fields = {
            "id": self.message_uuid,
            "session_id": self.session_id,
            "timestamp": self.created_at or datetime.utcnow().isoformat(),
            "content": content,
            "in_reply_to": self.in_reply_to,
            "metadata": self.message_metadata
        }
        
        # Add message type specific fields
        if self.type == MessageType.USER:
            return UserTextMessage(
                **base_fields,
                type=SocketMessageType.TEXT,
                from_user=True,
                to_agent=self.agent_id
            )
        elif self.type == MessageType.AGENT:
            return AgentTextMessage(
                **base_fields,
                type=SocketMessageType.TEXT,
                from_agent=self.agent_id,
                streaming=self.message_metadata.get("streaming", False) if self.message_metadata else False,
                turn_complete=self.message_metadata.get("turn_complete", True) if self.message_metadata else True
            )
        else:
            # Default to generic socket base message
            return SocketBaseMessage(
                **base_fields,
                type=SocketMessageType.SYSTEM
            )

class MessageCreate(BaseModel):
    """Schema for creating a new message."""
    type: MessageType
    role: Optional[MessageRole] = None
    agent_id: Optional[str] = None
    parts: List[Dict[str, Any]]
    message_metadata: Optional[Dict[str, Any]] = None
    in_reply_to: Optional[str] = None
    context_refs: Optional[List[Dict[str, Any]]] = None
    capabilities_used: Optional[List[str]] = None

class MessageRead(BaseModel):
    """Schema for reading message data."""
    id: int
    message_uuid: str
    session_id: str
    type: MessageType
    role: Optional[MessageRole]
    agent_id: Optional[str]
    parts: List[Dict[str, Any]]
    message_metadata: Optional[Dict[str, Any]]
    created_at: str
    updated_at: Optional[str]
    in_reply_to: Optional[str]
    context_refs: Optional[List[Dict[str, Any]]]
    capabilities_used: Optional[List[str]]

    class Config:
        from_attributes = True
    
    def to_socket_message(self) -> SocketMessage:
        """Convert read message to socket message format."""
        msg = Message(
            id=self.id,
            message_uuid=self.message_uuid,
            session_id=self.session_id,
            type=self.type,
            role=self.role,
            agent_id=self.agent_id,
            parts=self.parts,
            message_metadata=self.message_metadata,
            created_at=self.created_at,
            updated_at=self.updated_at,
            in_reply_to=self.in_reply_to,
            context_refs=self.context_refs,
            capabilities_used=self.capabilities_used
        )
        return msg.to_socket_message()

File: src/types/messages.ts
/**
 * Standardized Message Schema for TKR Multi-Agent Chat System
 * 
 * This file defines the canonical message formats used throughout the application
 * for communication between clients, agents, and the server.
 */

// Message Type Enum
export enum MessageType {
  TEXT = "text",                   // Plain text message
  AGENT_MESSAGE = "agent_message", // Agent-to-agent message
  SYSTEM = "system",               // System notification
  ERROR = "error",                 // Error message
  CONTEXT_UPDATE = "context_update", // Context sharing
  TASK_UPDATE = "task_update",     // Task status update
  PING = "ping",                   // Connection check
  PONG = "pong",                   // Connection check response
}

// Base Message Interface
export interface BaseMessage {
  // Core fields (required in all messages)
  id: string;                      // UUID for the message
  type: MessageType;               // Type of message
  timestamp: string;               // ISO-8601 formatted timestamp
  session_id: string;              // Session ID the message belongs to

  // Optional fields (may be present based on message type)
  from_agent?: string;             // ID of sending agent (null for user messages)
  to_agent?: string;               // ID of target agent (null for broadcast)
  from_user?: boolean;             // True if message is from a user 
  content?: string | object;       // Message content
  in_reply_to?: string;            // ID of message this is replying to
  
  // Streaming-related fields
  streaming?: boolean;             // True if message is part of a streaming response
  turn_complete?: boolean;         // True when streaming message is complete
  
  // Additional metadata
  metadata?: Record<string, any>;  // Additional message metadata
}

// User Text Message
export interface UserTextMessage extends BaseMessage {
  type: MessageType.TEXT;
  from_user: true;
  content: string;
  to_agent?: string;               // Optional target agent
}

// Agent Text Message
export interface AgentTextMessage extends BaseMessage {
  type: MessageType.TEXT;
  from_agent: string;
  content: string;
  streaming?: boolean;             // Whether this is a streaming message
  turn_complete?: boolean;         // Whether this completes the turn
}

// Agent-to-Agent Message
export interface AgentToAgentMessage extends BaseMessage {
  type: MessageType.AGENT_MESSAGE;
  from_agent: string;
  to_agent: string;
  content: any;                    // Structured content for A2A communication
  task_id?: string;                // Optional task ID this message is related to
}

// System Message
export interface SystemMessage extends BaseMessage {
  type: MessageType.SYSTEM;
  content: string;
  severity?: "info" | "warning";   // Optional severity for UI display
}

// Error Message
export interface ErrorMessage extends BaseMessage {
  type: MessageType.ERROR;
  content: string;
  error_code?: string;             // Error code for categorization
  error_details?: any;             // Additional error details
  severity?: "warning" | "error" | "critical"; // Error severity
  recoverable?: boolean;           // Whether the error is recoverable
}

// Context Update Message
export interface ContextUpdateMessage extends BaseMessage {
  type: MessageType.CONTEXT_UPDATE;
  from_agent: string;
  context_id: string;              // Unique ID for this context
  context_data: any;               // Context data being shared
  target_agents?: string[];        // Optional list of target agents (null = broadcast)
}

// Task Update Message
export interface TaskUpdateMessage extends BaseMessage {
  type: MessageType.TASK_UPDATE;
  task_id: string;                 // Task unique identifier
  status?: string;                 // Task status (pending, in_progress, completed, failed)
  action?: "create" | "update" | "cancel"; // Task action
  result?: any;                    // Optional task result data
}

// Connection Check Messages
export interface PingMessage extends BaseMessage {
  type: MessageType.PING;
}

export interface PongMessage extends BaseMessage {
  type: MessageType.PONG;
}

// Union type for all message types
export type Message = 
  | UserTextMessage
  | AgentTextMessage
  | AgentToAgentMessage
  | SystemMessage
  | ErrorMessage
  | ContextUpdateMessage
  | TaskUpdateMessage
  | PingMessage
  | PongMessage;

// Message parts for structured content
export interface MessagePart {
  type: string;
  content: any;
  metadata?: Record<string, any>;
}

// Socket.IO message status
export type MessageStatus = 
  | "sent"        // Message sent, not yet acknowledged
  | "delivered"   // Message delivered to recipient
  | "read"        // Message read by recipient
  | "error"       // Error during delivery
  | "pending"     // Message queued locally but not sent
  | "retrying";   // Message delivery being retried

// Response acknowledgment
export interface MessageAcknowledgment {
  status: MessageStatus;
  id: string;
  timestamp: string;
  error_message?: string;       // Present if status is "error"
  persisted_id?: string;        // ID in database storage
}